src/stake/stake_keys.rs

```diff
@@ -0,0 +1,170 @@
+    where
+        B: RangeBounds<Height>,
+    {
+        let bounds = normalize_range(bounds);
+
+        // Check if an equivalent stake intersects the arguments - if yes, extend it
+        if let Some(range) = self
+            .keys
+            .iter()
+            .filter(|(_, &staked)| staked == stake)
+            .find_map(|(range, _)| {
+                (range.contains(&bounds.start) || range.contains(&bounds.end)).then(|| range)
+            })
+        {
+            let range = range.to_owned();
+
+            let start = cmp::min(range.start, bounds.start);
+            let end = cmp::max(range.end, bounds.end);
+            let bounds = Range { start, end };
+
+            // Not using the entry manipulation because there is no inherent optimization in doing
+            // that for the cost of manually replacing the hash entry bucket of the key.
+            let _was_removed = self.keys.remove(&range).is_some();
+
+            debug_assert!(_was_removed);
+
+            self.keys.insert(bounds, stake);
+
+            return Ok(());
+        }
+
+        // Extend is not possible because either the key or value differs - intersection is forbidden
+        // from this point
+        if let Some(range) = self
+            .keys
+            .keys()
+            .find(|range| range.contains(&bounds.start) || range.contains(&bounds.end))
+        {
+            return Err(Error::DuplicatedStake {
+                height: range.start,
+                valid_before: range.end,
+            });
+        }
+
+        self.keys.insert(bounds, stake);
+
+        Ok(())
+    }
+
+    /// Remove all stake entries
+    pub(super) fn clear(&mut self) {
+        self.keys.clear();
+    }
+
+    /// Return a stake for a given height
+    pub(super) fn fetch(&self, height: Height) -> Option<&Stake> {
+        self.keys
+            .iter()
+            .filter_map(|(range, stake)| range.contains(&height).then(|| stake))
+            .next()
+    }
+
+    /// Remove all entries with the provided key
+    pub(super) fn purge_key(&mut self, key: &PublicKey) {
+        self.keys.retain(|_, stake| &stake.key != key);
+    }
+
+    /// Iter the ranges and stakes
+    pub(super) fn iter(&self) -> impl Iterator<Item = (&Range<Height>, &Stake)> {
+        self.keys.iter()
+    }
+}
+
+fn normalize_range<R>(bounds: R) -> Range<Height>
+where
+    R: RangeBounds<Height>,
+{
+    let start = match bounds.start_bound() {
+        Bound::Included(s) => *s,
+        Bound::Excluded(s) if s == &Height::MIN => Height::MIN,
+        Bound::Excluded(s) => s.saturating_add(1),
+        Bound::Unbounded => Height::MIN,
+    };
+
+    let end = match bounds.end_bound() {
+        Bound::Included(e) => e.saturating_add(1),
+        Bound::Excluded(e) => *e,
+        Bound::Unbounded => Height::MAX,
+    };
+
+    Range { start, end }
+}
+
+#[test]
+#[cfg(feature = "std")]
+fn stake_keys_intersect_and_merge() {
```

Comment:
Consider using rstest to parameterize and cleanup the code

Reply:
I'm not a bit fan of including external dependencies unless strictly needed. They not only increase the compilation time but introduce additional risk to the stack. In that case, we would need to increase the lines of code to include the appropriated fixtures to setup the dataset.

Reply:
While I understand the appeal of minimizing dependency hell across our crates, I wouldn't weigh dev-dependencies the same way. Things like rstest have no impact on downstream consumers of this library and only increase compile times for developers or CI pipelines doing uncached builds.